README.md
# Ribbon Maze

The maze itself is destined to be automated.

Ribbon Maze challenges you to explore a maze in search of rich resources. The maze is impassable, but new late game
technologies let you terraform the maze...

Mod portal page: https://mods.factorio.com/mod/RibbonMaze

## Key features

* **Perfect maze** - only one right path takes you deeper into a practically infinite maze. But...
* **Dead ends are rewarding!** - resources spawn at dead ends. Longer corridors hold the most valuable resources
* **Mixed ores near start** - mixed ores near the start of the maze provide an early game challenge, while
 making sure you can get your factory started
* **Great for trains** - resources are spread out so you'll soon want trains. The maze is "chunk aligned", ideal
for rail network blueprints
* **Mangroves** - the water at the start of the maze is lined with trees. Turn off normal trees in
 terrain generation if you wish. Mangroves offer wood as an infinite resource, like crude oil, but with new logistical
 challenges and opportunities
* **Terraforming** - Terraform the maze into canals using new recipes which provide a late-game need for
 wood and nuclear-generated steam. Automate the process with special artillery which will explore the maze for you.
 Decide: will you landfill your canals for more space, or exploit the new mangroves that grow along their banks?

## Soft modding

Ribbon Maze is designed to be "softmod" friendly. The core maze and resource generation logic can function
without the mangroves and terraforming prototypes. The mod's control.lua configures the settings, constants, and
events for the default mod experience, but all the logic is in separate scripts in the control directory. So
just providing a different control.lua should allow heavily customized setups, scenarios and soft mods.

## Credits

Terrain graphics for maze adapted from the MIT-license More Floors mod by Tone
(https://mods.factorio.com/mods/Tone/More_Floors).

Thanks to illiander42 for various modifications.

Thanks also to Factorio staff and forum users for their help and support!

changelog.txt
---------------------------------------------------------------------------------------------------
Version: 0.8.5
Date: 30. 11. 2020
  Bugfixes:
    - Updated to Factorio 1.1
---------------------------------------------------------------------------------------------------
Version: 0.8.4
Date: 28. 09. 2020
  Bugfixes:
    - Changed chasm color from green to black
---------------------------------------------------------------------------------------------------
Version: 0.8.3
Date: 05. 09. 2020
  Bugfixes:
    - Updated to Factorio 1.0
---------------------------------------------------------------------------------------------------
Version: 0.8.2
Date: 17. 06. 2020
  Bugfixes:
    - Migrating from 0.17 will no longer crash the game
---------------------------------------------------------------------------------------------------
Version: 0.8.1
Date: 23. 03. 2020
  Changes:
    - Ported to Factorio 0.18
---------------------------------------------------------------------------------------------------
Version: 0.7.6
Date: 11. 05. 2019
  Changes:
    - Minimum base version is now 0.17.38 to enforce fix to platform bit handling differences
  Bugfixes:
    - Had "tin" instead of "tin-ore" for bob's ores
---------------------------------------------------------------------------------------------------
Version: 0.7.5
Date: 08. 05. 2019
  Bugfixes:
    - Maze generation created lots of straight lines due to Linux/Windows fix in 0.7.4
---------------------------------------------------------------------------------------------------
Version: 0.7.4
Date: 06. 05. 2019
  Bugfixes:
    - Fixes multiplayer desync issue caused by numerical calculation differences between Windows and Linux
---------------------------------------------------------------------------------------------------
Version: 0.7.3
Date: 05. 05. 2019
  Bugfixes:
    - Migration added to fix maze terraforming artillery targeting for existing maps
---------------------------------------------------------------------------------------------------
Version: 0.7.2
Date: 04. 05. 2019
  Changes:
    - Updates to breaking changes in 0.17.35
    - Minimum base version is now 0.17.35
  Bugfixes:
    - Fix that in 0.17 maze terraforming artillery no longer target the maze chasm
---------------------------------------------------------------------------------------------------
Version: 0.7.1
Date: 25. 03. 2019
  Bugfixes:
    - Fish generation was using the wrong random number generator
---------------------------------------------------------------------------------------------------
Version: 0.7.0
Date: 16. 03. 2019
  Features:
    - Experimental support for factorio 0.17 alpha
  Balance:
    - Minimum dead end depth for Jivolite and Crotinnium reduced from 6 to 4 (angel's refining)
  Bugfixes:
    - Maximum clearing size was missing description in english locale
---------------------------------------------------------------------------------------------------
Version: 0.6.8
Date: 29. 01. 2019
  Features:
    - Merges options from illiander42's variant
    - Fish added
    - Loop chance option: Now your maze can have loops in it. Beware that adding loops reduces the number of dead ends, which means less resource patches.
    - Clearings option: gives a chance for large open areas to spawn. These will almost certainly create loops even without the imperfect maze option.
    - Oasis: Option to find mangrove-filled water at dead-ends, instead of ore or oil.
    - Settings to control mangrove density and proportion of raw/green wood
  Balance:
    - Default mangrove ratio of green to raw wood is higher
---------------------------------------------------------------------------------------------------
Version: 0.6.7
Date: 07. 12. 2018
  Major Features:
    - Mod ores can now be put at dead ends like vanilla ores. Supports practically any mod, with manual fine-tuning for popular mods such as
      Bob's Ores, Angel's Refining, Angel's Infinite Ores, Yuoki Industries, and more. Enable with the "Dead end mod resources" setting.
      (https://github.com/h8ul-modder/factorio-ribbon-maze/issues/1)
    - Maze block size can now be customized. Allowed sizes: 32 (default), 64, 96, 128, 192, 256. https://github.com/h8ul-modder/factorio-ribbon-maze/issues/22
  Features:
    - Settings to control how infinite ores from mods are handled. (https://github.com/h8ul-modder/factorio-ribbon-maze/issues/20)
    - Setting to stretch out resource richness increase. (https://github.com/h8ul-modder/factorio-ribbon-maze/issues/21)
    - Further improved spacing for oil and other resources with multi-tile collision boxes. (https://github.com/h8ul-modder/factorio-ribbon-maze/issues/27)
    - Setting to have mixed ores throughout the maze (https://github.com/h8ul-modder/factorio-ribbon-maze/issues/28)
  Balance:
    - Reduce frequency of Mangrove (Bruguiera) by half. (https://github.com/h8ul-modder/factorio-ribbon-maze/issues/17)
    - Less individual oil entities now that the minimum yield of 20% is respected.
  Bugfixes:
    - Mangrove (Bruguiera) was not being removed in terraforming. (https://github.com/h8ul-modder/factorio-ribbon-maze/issues/16)
    - Maze terraforming collision mask fix. (https://github.com/h8ul-modder/factorio-ribbon-maze/issues/23)
    - Oil spacing changed in existing saves. NOTE: you'll need to MOVE YOUR PUMPJACKS! (https://github.com/h8ul-modder/factorio-ribbon-maze/issues/25)
  Optimisations:
    - Multi-tile resource alignment is now calculated at configuration time rather than at chunk generation.
---------------------------------------------------------------------------------------------------
Version: 0.6.6
Date: 30. 11. 2018
  Features:
    - New setting: Patchwork Size for Mixed Resources. (https://github.com/h8ul-modder/factorio-ribbon-maze/issues/11)
    - Grid-based crude oil entity placement with better spacing (https://github.com/h8ul-modder/factorio-ribbon-maze/issues/10)
  Optimisations:

settings.lua
--[[
   Copyright 2018 H8UL

   Permission is hereby granted, free of charge, to any person obtaining a copy
   of this software and associated documentation files (the "Software"), to deal
   in the Software without restriction, including without limitation the rights
   to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
   copies of the Software, and to permit persons to whom the Software is
   furnished to do so, subject to the following conditions:

   The above copyright notice and this permission notice shall be included in all
   copies or substantial portions of the Software.

   THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
   IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
   FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
   AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
   LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
   OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
   SOFTWARE.
--]]

data:extend{
    {
        type = "int-setting",
        name = "ribbon-maze-blocks",
        setting_type = "runtime-global",
        min_value = 1,
        max_value = 70,
        default_value = 21,
        order="rb-a",
    },
    {
        type = "int-setting",
        name = "ribbon-maze-block-size",
        setting_type = "runtime-global",
        default_value = 32,
        allowed_values = {32,64,96,128,192,256},
        order="rb-b",
    },
    {
        type = "int-setting",
        name = "ribbon-maze-clear-start",
        setting_type = "runtime-global",
        default_value = 0,
        order="rb-c",
    },
    {
        type = "bool-setting",
        name = "ribbon-maze-chart-nearby-crude-oil",
        setting_type = "runtime-global",
        default_value = false
        ,
        order="rb-d",
    },
    {
        type = "bool-setting",
        name = "ribbon-maze-chart-nearby-uranium-ore",
        setting_type = "runtime-global",
        default_value = false,
        order="rb-e",
    },
    {
        type = "bool-setting",
        name = "ribbon-maze-mixed-preferred",
        setting_type = "runtime-global",
        default_value = false,
        order="rb-f-a",
    },
    {
        type = "int-setting",
        name = "ribbon-maze-mixed-patchwork-min",
        setting_type = "runtime-global",
        default_value = 1,
        allowed_values = {1,2,3,5,6,10,15},
        order="rb-f-b",
    },
    {
        type = "int-setting",
        name = "ribbon-maze-mixed-patchwork-max",
        setting_type = "runtime-global",
        default_value = 15,
        allowed_values = {1,2,3,5,6,10,15},
        order="rb-f-c",
    },
    {
        type = "double-setting",
        name = "ribbon-maze-resource-stretch-factor",
        setting_type = "runtime-global",
        minimum_value = 0.01,
        maximum_value = 1000,
        default_value = 1,
        order="rb-resource-stretch-factor",
    },
    {
        type = "bool-setting",
        name = "ribbon-maze-stone",
        setting_type = "runtime-global",
        default_value = true,
        order="rb-h",
    },
    {
        type = "bool-setting",
        name = "ribbon-maze-coal",
        setting_type = "runtime-global",
        default_value = true,
        order="rb-i",
    },
    {
        type = "bool-setting",
        name = "ribbon-maze-copper-ore",
        setting_type = "runtime-global",
        default_value = true,
        order="rb-j",
    },
    {
        type = "bool-setting",
        name = "ribbon-maze-iron-ore",
        setting_type = "runtime-global",
        default_value = true,
        order="rb-k",
    },
    {
        type = "bool-setting",
        name = "ribbon-maze-uranium-ore",
        setting_type = "runtime-global",
        default_value = true,
        order="rb-lu",
    },
    {
        type = "bool-setting",
        name = "ribbon-maze-crude-oil",
        setting_type = "runtime-global",
        default_value = true,
        order="rb-lc",
    },
    {
        type = "bool-setting",
        name = "ribbon-maze-mod-resources",
        setting_type = "runtime-global",
        default_value = false,
        order="rb-m",
    },
    {
        type = "bool-setting",
        name = "ribbon-maze-separate-out-infinite-ores",
        setting_type = "runtime-global",
        default_value = false,
        order="rb-m-infinite",
    },
    {
        type = "double-setting",
        name = "ribbon-maze-infinite-ore-stretch-factor",
        setting_type = "runtime-global",
        minimum_value = 0.01,
        maximum_value = 1000,
        default_value = 8,
        order="rb-m-infinite-stretch-factor",
    },
    {
        type = "double-setting",
        name = "ribbon-maze-mangrove-density",
        setting_type = "runtime-global",
        minimum_value = 0.01,
        maximum_value = 1,
        default_value = 0.5,
        order="rb-n-density",
    },
    {
        type = "double-setting",
        name = "ribbon-maze-green-raw-ratio",
        setting_type = "runtime-global",
        minimum_value = 0.5,
        maximum_value = 1.0,
        default_value = 0.95,
        order="rb-n-density",
    },
    {
        type = "double-setting",
        name = "ribbon-maze-loop-chance",
        setting_type = "runtime-global",
        minimum_value = 0,
        maximum_value = 1,
        default_value = 0,
        order="rb-v-a",
    },
    {
        type = "bool-setting",
        name = "ribbon-maze-water-resource",
        setting_type = "runtime-global",
        default_value = false,
        order="rb-v-b",
    },
    {
        type = "double-setting",
        name = "ribbon-maze-clearing-chance",
        setting_type = "runtime-global",
        minimum_value = 0,
        maximum_value = 1,
        default_value = 0,
        order="rb-v-c-c",
    },
    {
        type = "int-setting",
        name = "ribbon-maze-clearing-size-max",
        setting_type = "runtime-global",
        default_value = 3,
        order="rb-v-c-m",
    },
}

